Add RNA-seq Ortholog Mapping example notebook (#267)
AlexsLemonade/refinebio-examples: 03-rnaseq/ortholog-mapping_rnaseq_01_ensembl.Rmd

Fill out the pop up window with your email and our Terms and Conditions:  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/e140face75daa6d2c34e30a4755c362e6039a677/template/screenshots/download-email.png" width=500>  

We are going to use non-quantile normalized data for this analysis.
To get this data, you will need to check the box that says "Skip quantile normalization for RNA-seq samples".
Note that this option will only be available for RNA-seq datasets.

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/e140face75daa6d2c34e30a4755c362e6039a677/template/screenshots/skip-quantile-normalization.png" width=500>

It may take a few minutes for the dataset to process.
You will get an email when it is ready. 

## About the dataset we are using for this example

For this example analysis, we will use this [acute myeloid leukemia sample dataset](https://www.refine.bio/experiments/SRP070849/combination-targeted-therapy-to-disrupt-aberrant-oncogenic-signaling-and-reverse-epigenetic-dysfunction-in-idh2-and-tet2-mutant-acute-myeloid-leukemia-rna-seq).

> The data that we downloaded from refine.bio for this analysis has 19 samples (obtained from 19 mice with acute myeloid leukemia (AML)), containing RNA-sequencing results for types of AML under controlled treatment conditions.

Reviewer: This probably needs more detail. I don't think the mice actually have AML, I think they are using different mouse models of AML and they also did some drug treatments on the different mouse models. 

Even though we aren't using this information for the mapping per se, we should still make sure the descriptions are on point.

)) %>%
  # Tuck away the gene ID column as rownames
  tibble::column_to_rownames("Gene")
```

Let's ensure that the metadata and data are in the same sample order. 

```{r}
# Make the data in the order of the metadata
df <- df %>%
  dplyr::select(metadata$refinebio_accession_code)

# Check if this is in the same order
all.equal(colnames(df), metadata$refinebio_accession_code)

# Bring back the "Gene" column in preparation for mapping
df <- df %>%
  tibble::rownames_to_column("Gene")
```

## Mapping mouse Ensembl gene IDs to human symbols

refine.bio data uses Ensembl gene identifiers, which will be in the first column.

```{r}
# Let's take a look at the first 6 items in the `Gene` column of `df`
head(df$Gene)
```

Ensembl gene identifiers have different species-specific prefixes. 
In mouse, Ensembl gene identifiers begin with `ENSMUSG` (in human, `ENSG`, etc.).

Now let's do the conversion!
We'll use the `mouse` data from `hcop`; this is what `mouse` looks like.

```{r}
mouse
```

We're interested in the `human_symbol` and `mouse_ensembl` columns specifically.

```{r}
human_mouse_key <- mouse %>%
  # we'll want to subset mouse to only the columns we're interested in

>   dplyr::select(mouse_ensembl, human_symbol)

Reviewer: Do you need to do the same `distinct()` steps you did in microarray? My guess is you do, but I haven't ran this myself yet.

# Now let's join the expression data from `df` for the subset of genes in `df`
human_mouse_mapped_df <- df %>%
  dplyr::left_join(human_mouse_key, by = c("Gene" = "mouse_ensembl"))
```

Here's what the new data frame looks like:

```{r}
head(human_mouse_mapped_df, n = 25)
```

Let's get a summary of the gene symbols returned in the `human_symbol` column of our mapped data frame.

```{r}
# We can use the `count()` function to get a count of the distribution of values in the `human_symbol` column
human_mouse_mapped_df %>%
  dplyr::group_by(human_symbol) %>%
  dplyr::count() %>%
  # Sort by highest `n` which would be the human gene symbol with the most mapped mouse Ensembl IDs
  dplyr::arrange(desc(n))
```

Looks like we have mapped symbols!
We also have 18908 NAs, which is okay, because we do not expect everything to map across species.

## Take a look at some multi-mappings

If a mouse Ensembl gene ID maps to multiple human symbols, the gene expression values get duplicated.
Let's look at the `ENSMUSG00000000037` example below.

```{r}
human_mouse_mapped_df %>%
  dplyr::filter(Gene == "ENSMUSG00000000037") %>%

>   dplyr::select(Gene, human_symbol, dplyr::everything())

Reviewer: I would say the same comment here as I did in the microarray PR, we should reorder this once in an earlier step so we don't have to do this twice.

To get this data, you will need to check the box that says "Skip quantile normalization for RNA-seq samples".
Note that this option will only be available for RNA-seq datasets.

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/e140face75daa6d2c34e30a4755c362e6039a677/template/screenshots/skip-quantile-normalization.png" width=500>

It may take a few minutes for the dataset to process.
You will get an email when it is ready. 

## About the dataset we are using for this example

For this example analysis, we will use this [acute myeloid leukemia sample dataset](https://www.refine.bio/experiments/SRP070849/combination-targeted-therapy-to-disrupt-aberrant-oncogenic-signaling-and-reverse-epigenetic-dysfunction-in-idh2-and-tet2-mutant-acute-myeloid-leukemia-rna-seq).

The data that we downloaded from refine.bio for this analysis has 19 samples (obtained from 19 mouse models with acute myeloid leukemia (AML)), containing RNA-sequencing results for types of AML under controlled treatment conditions.
More specifically, IDH2-mutant AML mouse models were treated with either vehicle or AG-221 (the first small molecule in-vivo inhibitor of IDH2 to enter clinical trials).

> While, the TET2-mutant AML mouse models were treated with vehicle or 5-Azacytidine (Decitabine, hypomethylating agent).

Author: They are actually two different drugs (source: https://www.clinicaltrials.gov/ct2/show/NCT03383575#armgroup)

The `support` column contains a list of associated databases that support each assertion.
This column may assist with addressing some of the multi-mappings that we will talk about later.

```{r}
human_mouse_key <- mouse %>%
  # We'll want to subset mouse to only the columns we're interested in
  dplyr::select(mouse_ensembl, human_ensembl, support)

# Since we ignored the additional columns in `mouse`, let's check to see if we have any duplicates in our `human_mouse_key`
any(duplicated(human_mouse_key))
```

We do have duplicates!
We don't want to handle duplicate data, so let's remove those duplicates before moving forward.

```{r}
human_mouse_key <- human_mouse_key %>%
  # We need to use the `distinct()` function to remove duplicates resulted from ignoring the additional columns in the `mouse` object
  dplyr::distinct()
```

Now let's join the mapped data from `human_mouse_key` with the expression data in `df`.

```{r}
human_mouse_mapped_df <- df %>%
  dplyr::left_join(human_mouse_key, by = c("Gene" = "mouse_ensembl")) %>%
  # Sort so that the columns with the mouse Ensembl IDs and human Ensembl IDs are first

>   dplyr::select(Gene, human_ensembl, dplyr::everything())

Reviewer: Can we bring the `support` column to the front too? I forgot it was there.

human_mouse_mapped_df %>%
  dplyr::group_by(human_ensembl) %>%
  dplyr::count() %>%
  # Sort by highest `n` which would be the human Ensembl ID with the most mapped mouse Ensembl IDs
  dplyr::arrange(desc(n))
```

Looks like we have mapped IDs!
Now, let's get an idea of how many mouse Ensembl IDs we have that were not mapped to human Ensembl IDs.

```{r}
sum(is.na(human_mouse_mapped_df$human_ensembl))
```

We have 18801 NAs, which means we have 18801 mouse Ensembl IDs that were not mapped to human Ensembl IDs.
This is okay because we do not expect everything to map across species.

## Take a look at some multi-mappings

If a mouse Ensembl gene ID maps to multiple human Ensembl IDs, the gene expression values get duplicated.
Let's look at the `ENSMUSG00000000290` example below.

```{r}
human_mouse_mapped_df %>%
  dplyr::filter(Gene == "ENSMUSG00000000290")
```

Remember that our original data is associated to mouse Ensembl IDs, so when a human Ensembl ID maps to multiple mouse Ensembl IDs, the data will not get duplicated, but you will have multiple rows associated with that human symbol.
Let's look at the `CACNA1B` example below.

```{r}
human_mouse_mapped_df %>%

>   dplyr::filter(human_ensembl == "CACNA1B")

Reviewer: This step produces 0 rows. I think you need to update it.

```{r}
collapsed_human_ensembl_df <- human_mouse_mapped_df %>%
  # Group by Ensembl IDs
  dplyr::group_by(Gene) %>%
  # Collapse the mapped values in `human_mouse_mapped_df` into one column named `all_human_ensembls`
  dplyr::summarize(all_human_ensembls = paste(human_ensembl, collapse = ";")) %>%
  # Now we need get the gene expression data back
  dplyr::left_join(df, by = "Gene")

head(collapsed_human_ensembl_df)
```

### Write results to file

Now let's write our results to file.

```{r}
readr::write_tsv(
  collapsed_human_ensembl_df,
  file.path(
    results_dir,
    "SRP070849_mouse_ensembl_to_collapsed_human_gene_symbol.tsv" # Replace with a relevant output file name
  )
)
```

## Collapse multi-mapped mouse Ensembl IDs

Since multiple mouse Ensembl gene IDs map to the same human symbol, we can also collapse all Ensembl IDs that each unique human Ensembl ID mapped to.

A known issue in ortholog mapping is how to handle this data because combining values across mouse transcripts using an average may result in the loss of a lot of data and will likely not be representative of the mouse biology.

> Therefore, we will not add the expression data here since the expression values are not the same across Ensembl IDs, despite mapping to the same human Ensembl ID.

Author: This was actually an outdated comment, removed it in the last commit.

Instead, we will use the `support` column to decide which mappings to retain.
As we noted earlier, the prediction with the most number of databases which concur is considered the most reliable.
Therefore, where we have multi-mapped mouse Ensembl gene IDs, we will take the mapping with the most databases that concur according to the `support` column.

```{r}

> filtered_mouse_ensembl_df <- human_mouse_mapped_df %>%

Reviewer: I would split this up a bit (show them a preview of the support column) and print out things more so we can show what's happening at each step.

  dplyr::mutate(n_databases = stringr::str_count(support, ",") + 1) %>% # Count the number of databases in the support column for each prediction
  dplyr::group_by(human_ensembl) %>% # Group by human Ensembl IDs
  dplyr::filter(n_databases == max(n_databases)) # Now filter for the row with the largest number of databases in support for each human Ensembl ID

head(filtered_mouse_ensembl_df)
```

### Write results to file

Now let's write our list of collapsed mouse Ensembl IDs for each unique human Ensembl ID to file.

```{r}
readr::write_tsv(
  filtered_mouse_ensembl_df,
  file.path(
    results_dir,

>     "SRP070849_filtered_mouse_ensembl_to_human_gene_symbol_list.tsv" # Replace with a relevant output file name

Reviewer: ```suggestion
    "SRP070849_filtered_mouse_ensembl_to_human_gene_symbol.tsv" # Replace with a relevant output file name
```
We don't wanna be using the word `list` when it isn't a list.

We suggest saving plots and results to `plots/` and `results/` directories, respectively, as these are automatically created by the notebook.
From here you can customize this analysis example to fit your own scientific questions and preferences. 

***

<!-- Do not delete this line --> <a name="analysis" style="padding-top:56px;margin-top:-56px;">&nbsp;</a>

# Ortholog Mapping - RNA-seq

## Install libraries

See our Getting Started page with [instructions for package installation](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#what-you-need-to-install) for a list of the software you will need, as well as more tips and resources.

Attach a package we need for this analysis.

```{r}
# We will need this so we can use the pipe: %>%
library(magrittr)
```

## Import data from HGNC

The HUGO Gene Nomenclature Committee (HGNC) assigns a unique and ideally meaningful name and symbol to every human gene.

> The [HGNC database](https://www.genenames.org/) currently contains over 39000 public records containing approved human gene nomenclature and associated gene information [@Gray2015].

Reviewer: ```suggestion
The [HGNC database](https://www.genenames.org/) currently contains over 39,000 public records containing approved human gene nomenclature and associated gene information [@Gray2015].
```
I find numbers easier to read with commas.

The [HGNC Comparison of Orthology Predictions (HCOP)](https://www.genenames.org/tools/hcop/) is a search tool that combines orthology predictions for a specified human gene, or set of human genes from a variety of sources, including Ensembl Compara, HGNC, and NCBI Gene Orthology.

> Thus allowing users to obtain and compare orthology data [@Wright2005].

Reviewer: This is not a complete sentence (Was this in the microarray example too?)

Author: Good catch, I'll check the microarray example and make the change there in this PR as well (if it it's in there).

In general, an orthology prediction where most of the databases concur would be considered the reliable, and we will use this to prioritize mapping in cases where there is more than one possible ortholog for a gene.
HCOP was originally designed to show orthology predictions between human and mouse, but has been expanded to include data from 18 genomes [@hcop-help].

> First, we need to download the file from the FTP server holding the HGNC data.

Reviewer: I don't imagine FTP means much to our users. ¯\\_(ツ)_/¯ and I don't think we need to explain. 
```suggestion
First, we need to download the file from the server holding the HGNC data.
```

This is where the files that reflect the data provided via the [HGNC database](https://www.genenames.org/) are maintained.
Ortholog species files with the '6 Column' output returns the raw assertions, Ensembl gene IDs and Entrez Gene IDs for human and one other species, while the '15 Column' output includes additional information such as the chromosomal location, accession numbers and the databases that support the assertions.

*Note:* If you are using Safari (or the above FTP server link does not open in a web browser), you may need to go to the [link for the HCOP search tool](https://www.genenames.org/tools/hcop/) and scroll down to "Bulk Downloads" to choose a file to download.
Here, you can find the same files you would find at the FTP server linked above.

To download a file, click the file name.
For this notebook, you will want to download the file named `human_mouse_hcop_fifteen_column.txt.gz`.
If you are using a different dataset, you can replace `mouse` in `human_mouse_hcop_fifteen_column.txt.gz` with the name of the species you have data for, and click on that file to download.

> <img src="https://github.com/AlexsLemonade/refinebio-examples/raw/3d4e8a9be91eee5e95a5875d5e897e9ed4cfeac4/template/screenshots/download-hgnc-mouse-file.png" width=400>

Reviewer: I'm debating about whether we should keep all images as the same commit id, but I guess this is the only place this image is used so maybe its fine.

Next, move the `human_mouse_hcop_fifteen_column.txt.gz` file into your `data/` folder.

*Note:* If you are using Safari, this file will automatically be decompressed, so the name of the file would instead be `human_mouse_hcop_fifteen_column.txt` (don't forget to change the file name in the chunk below if this is the case).

In the next chunk, we will read in the orthology file that was just downloaded.

```{r}
# Define the file path to organism orthology file downloaded from the HGNC database
mouse_hgnc_file <- file.path("data", "human_mouse_hcop_fifteen_column.txt.gz")

> ```

Reviewer: Could do or encourage a `file.exists()`.

Reviewer: But I'm not 100% sold on this idea.

In the next chunk, we show how we can collapse all the human Ensembl IDs into one column where we store them all for each unique mouse Ensembl ID.

```{r}
collapsed_human_ensembl_df <- human_mouse_mapped_df %>%
  # Group by mouse Ensembl IDs
  dplyr::group_by(Gene) %>%
  # Collapse the mapped values in `human_mouse_mapped_df` into one column named
  # `all_human_ensembls` -- note that we will lose the `support` column in this summarizing step
  dplyr::summarize(all_human_ensembls = paste(human_ensembl, collapse = ";"))

head(collapsed_human_ensembl_df)
```

### Write results to file

Now let's write our results to file.

```{r}
readr::write_tsv(
  collapsed_human_ensembl_df,
  file.path(
    results_dir,
    # Replace with a relevant output file name
    "SRP070849_mouse_ensembl_to_collapsed_human_gene_symbol.tsv"
  )
)
```

## Collapse human genes mapping to multiple mouse genes

Since multiple mouse Ensembl gene IDs map to the same human Ensembl gene ID, we may want to identify which one of these mappings represents the "true" ortholog, i.e. which mouse gene is most similar to the human gene we are interested in.

> *This is not at all straightforward!* (see, for just one example, https://doi.org/10.1093/bioinformatics/btaa468) [@Stamboulian2020].

Reviewer: Probably should stick to our general format for links. 
```suggestion
*This is not at all straightforward!* (see, for just one example [this paper](https://doi.org/10.1093/bioinformatics/btaa468)) [@Stamboulian2020].
```

Reviewer 2: *This is not at all straightforward!* (See [this paper](https://doi.org/10.1093/bioinformatics/btaa468) for just one example [@Stamboulian2020].)
